File: FrontEnd/lib/domain/entities/post.dart
/// @file post.dart
/// @description Entidad de dominio Post. Representa una publicación en el feed de la comunidad.
/// Objeto puro de Dart, sin dependencias de Flutter ni paquetes externos.
/// Los mappers convierten PostModel ↔ Post.
/// @module Community
/// @layer Domain
library;

// ═══════════════════════════════════════════════════════════════════════════════
// POST ENTITY
// ═══════════════════════════════════════════════════════════════════════════════

/// Entidad de dominio que representa un post de la comunidad.
///
/// Incluye datos del autor embebidos (authorName, authorPhoto) para evitar
/// llamadas adicionales al cargar el feed (desnormalización de lectura).
/// Todos los campos son inmutables (final). Usar [copyWith] para actualizaciones.
class Post {
  const Post({
    required this.id,
    required this.userId,
    required this.authorName,
    required this.content,
    required this.likesCount,
    required this.commentsCount,
    required this.createdAt,
    required this.updatedAt,
    required this.isLikedByMe,
    this.authorPhoto,
    this.image,
  });

  /// Identificador único del post (MongoDB ObjectId como String).
  final String id;

  /// ID del usuario autor del post.
  final String userId;

  /// Nombre visible del autor (embebido para rendimiento en feed).
  final String authorName;

  /// URL de la foto de perfil del autor (Cloudinary), o null si no tiene.
  final String? authorPhoto;

  /// Contenido textual del post (máx. 1000 caracteres).
  final String content;

  /// URL de la imagen adjunta al post (Cloudinary), o null si no tiene.
  final String? image;

  /// true si el usuario autenticado ya dio like a este post.
  final bool isLikedByMe;

  /// Número de likes acumulados (desnormalizado en base de datos).
  final int likesCount;

  /// Número de comentarios acumulados (desnormalizado en base de datos).
  final int commentsCount;

  /// Fecha UTC de creación del post.
  final DateTime createdAt;

  /// Fecha UTC de la última actualización.
  final DateTime updatedAt;

  // ─── Helpers ─────────────────────────────────────────────────────────────────

  /// true si el post tiene imagen adjunta.
  bool get hasImage => image != null && image!.isNotEmpty;

  /// true si el autor tiene foto de perfil.
  bool get hasAuthorPhoto => authorPhoto != null && authorPhoto!.isNotEmpty;

  /// true si el post tiene al menos un like.
  bool get hasLikes => likesCount > 0;

  /// true si el post tiene al menos un comentario.
  bool get hasComments => commentsCount > 0;

  // ─── copyWith ────────────────────────────────────────────────────────────────

  /// Devuelve una copia del post con los campos indicados modificados.
  /// Útil para actualizaciones optimistas de likesCount y commentsCount.
  Post copyWith({
    String?   id,
    String?   userId,
    String?   authorName,
    String?   authorPhoto,
    String?   content,
    String?   image,
    int?      likesCount,
    int?      commentsCount,
    bool?     isLikedByMe,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Post(
      id:            id            ?? this.id,
      userId:        userId        ?? this.userId,
      authorName:    authorName    ?? this.authorName,
      authorPhoto:   authorPhoto   ?? this.authorPhoto,
      content:       content       ?? this.content,
      image:         image         ?? this.image,
      likesCount:    likesCount    ?? this.likesCount,
      commentsCount: commentsCount ?? this.commentsCount,
      isLikedByMe:   isLikedByMe   ?? this.isLikedByMe,
      createdAt:     createdAt     ?? this.createdAt,
      updatedAt:     updatedAt     ?? this.updatedAt,
    );
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) || (other is Post && other.id == id);

  @override
  int get hashCode => id.hashCode;

  @override
  String toString() => 'Post(id: $id, userId: $userId, likesCount: $likesCount)';
}

File: BackEnd/src/domain/entities/Post.ts
/**
 * @file Post.ts
 * @description Entidad de dominio que representa un post de la comunidad.
 * Objeto puro: sin dependencias de framework ni de base de datos.
 * @module Community
 * @layer Domain
 */

/**
 * Entidad de dominio Post.
 *
 * Representa una publicación en el feed de la comunidad.
 * Los contadores `likesCount` y `commentsCount` están desnormalizados
 * en el documento de MongoDB para evitar COUNT queries en el feed.
 */
export class Post {
  /** Identificador único (ObjectId de MongoDB serializado como string) */
  readonly id: string;

  /** ID del usuario autor del post */
  readonly userId: string;

  /** Contenido textual del post (máx. 1000 caracteres) */
  readonly content: string;

  /** URL de la imagen en Cloudinary, si se adjuntó una */
  readonly image?: string;

  /** Número de likes acumulados (desnormalizado para rendimiento) */
  readonly likesCount: number;

  /** Número de comentarios acumulados (desnormalizado para rendimiento) */
  readonly commentsCount: number;

  /** IDs de usuarios que han visto el post (para métricas de alcance) */
  readonly seenBy: string[];

  /** Fecha de creación del post */
  readonly createdAt: Date;

  /** Fecha de la última actualización */
  readonly updatedAt: Date;

  /** Fecha de borrado lógico. Si está definida, el post está eliminado */
  readonly deletedAt?: Date | null;

  constructor(params: {
    id: string;
    userId: string;
    content: string;
    image?: string;
    likesCount?: number;
    commentsCount?: number;
    seenBy?: string[];
    createdAt: Date;
    updatedAt: Date;
    deletedAt?: Date | null;
  }) {
    this.id            = params.id;
    this.userId        = params.userId;
    this.content       = params.content;
    this.image         = params.image;
    this.likesCount    = params.likesCount    ?? 0;
    this.commentsCount = params.commentsCount ?? 0;
    this.seenBy        = params.seenBy        ?? [];
    this.createdAt     = params.createdAt;
    this.updatedAt     = params.updatedAt;
    this.deletedAt     = params.deletedAt;
  }

  /** true si el post no ha sido eliminado lógicamente */
  get isActive(): boolean {
    return !this.deletedAt;
  }
}
